Fiber Reconciler fixes

ahti opened the pull request:
This is what I have done so far to make the example (with a little modification) from #523 work.

Currently, things work until returning to the initial state, after which pressing the button will leave an empty screen.

Some comments in no particular order:

* The reproducer target is of course only for debugging and to be removed before merge.
* I've left the logs as normal prints for now, since [swift-log doesn't build on wasm rn](https://github.com/apple/swift-log/pull/236), we can remove those as well if things get ready before swift-log is fixed.
* Pay no attention to the individual commits, I'll clean those up once things look more mergeable.
* The implementation of `ReconcilePass.reconcile(_: , in:, caches:)` obviously still needs to be refactored.
* The reconciliation pass currently emits mutations that seem to assume replacing an element transfers any children to the replacement. I've adjusted the DOM and test renderers, but maybe the emitted mutations should be changed instead.
* I tried turning the reproducer into a test case, but that fails one step earlier, while returning to the initial state, due to an out of range index in an insert mutation.
* Another issue I observed: There are cases where a `remove` mutation is emitted with the wrong `parent`. The `DOMRenderer` doesn't use the parent, so it's not an issue there, but it does materialize with the test renderer.
* The parent in the remove mutation is optional, but all places where such a value is created pass a non-nil parent. Is there a reason for this, or can this be turned non-optional?

ahti committed:
prints n stuff

ahti committed:
reproducer

ahti committed:
wip

ahti committed:
wip

ahti committed:
fixed!

ahti committed:
whoops new issue

ahti committed:
wip

carson-katri commented:
I don't recall specifically, but the `parent` may be be optional because the root of the hierarchy has no parent and could hypothetically be removed.

carson-katri commented:
Tested on a small project and didn't notice any regressions.

ahti committed:
Merge main into investigate

ahti committed:
check node.fiber in the beginning

ahti committed:
refactor ReconcilePass.reconcile(...)

ahti committed:
fix going from state a -> b after going around once in RecRep

ahti commented:
I finally found some time and started refactoring `ReconcilePass.reconcile` (not quite done with that), and also fixed the empty screen when going through all the states of the reproducer and then from `.a` to `.b` again (nothing was inserted when a fiber that should have an element had an alternate, but the alternate didn't have an element).

There is still an issue with label content not being updated tho, which I'm going to look into next.

ahti committed:
update reinserted element with newContent

ahti committed:
update and extend test

ahti commented:
Fixed the stale label content as well.

The only issue left that I am aware of is the `.remove` mutation with wrong parent. I've extended the test to reproduce that one, too.

I also changed the test renderer so that multiple changed fibers get batched like they do in real renderers. Without that, the test was kind of flaky and would randomly succeed or fail (before adding the part to reproduce the remove bug, that is).

carson-katri commented on Package.swift:
```diff
@@ -15,6 +15,7 @@ let package = Package(
       name: "TokamakDemo",
       targets: ["TokamakDemo"]
     ),
+    .executable(name: "RecRep", targets: ["RecRep"]),
```
I'm not sure we should have this executable as part of the Tokamak package. It seems you have already adapted it to a test case, which should be enough.

carson-katri commented on Sources/TokamakCore/Fiber/FiberReconciler.swift:
```diff
@@ -270,6 +270,14 @@ public final class FiberReconciler<Renderer: FiberRenderer> {
     self.alternate = current
     current = alternate
 
+    print("""
+      reconcile done
+      mutations were:
+      \(visitor.mutations.map { "  \($0)" }.joined(separator: "\n"))
+      alternate is \(self.alternate.recursiveDescription)
+      current is \(current.recursiveDescription)
+      """)
```
Can you clean up any print statements before merging? Better debug logging could be tackled in a separate PR.

carson-katri commented on Sources/TokamakCore/Fiber/Passes/ReconcilePass.swift:
```diff
@@ -223,41 +231,73 @@ struct ReconcilePass: FiberReconcilerPass {
+
+    invalidateCache(for: fiber, in: reconciler, caches: caches)
+
+    switch (fiber.element, node.newContent, fiber.alternate?.element) {
```
Very nice improvement here!

carson-katri commented on Sources/TokamakCore/Fiber/Passes/ReconcilePass.swift:
```diff
@@ -223,41 +231,73 @@ struct ReconcilePass: FiberReconcilerPass {
+      if let p = fiber.elementParent { caches.elementIndices[ObjectIdentifier(p)]? -= 1 }
+      return .remove(element: alt, parent: parent)
+
+    case let (element?, content, _) where fiber.alternate?.element == nil:
```
Couldn't this be written as:
```suggestion
    case let (element?, content, nil):
```

carson-katri reviewed (changes requested):
Very nice improvements, thank you. I used it on a test project of mine and it worked well.

ahti committed:
fix wrong parent in remove mutation

ahti committed:
clean up reproducer target

ahti committed:
undo logging changes

ahti committed:
simplify case pattern

ahti commented:
superseeded by #542
